### PR TITLE
Introduce error classes in an attempts to better handle errors in the library

### PR DESCRIPTION
**Pull request by ddolcimascolo**

### CODE REVIEW

File: lib/client/Client.js [hunk 1]
```diff
@@ -655,22 +656,24 @@ export default class Client {
...
+      .post(this.apiUrl, this._defaultHeaders(), [[method, options || {}, '#0']])
+      .then(data => Utils.assertValidJMAPResponse(method, data))
+      .then(data => data.map(response => this._handleResponse(response, method)))
+      .then(responses => responses.length > 1 ? responses : responses[0]);
   }
 
-  _handleResponse(response) {
+  _handleResponse(response, method) {
     let name = response[0],
         fn = this['_handle' + Utils.capitalize(name) + 'Response'];
```
Comment (dbenchi) [inline lib/client/Client.js]: Not related to your PR, but what do you think about backtick here?

File: lib/errors/JmapError.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  constructor(payload, method) {
+    super();
+
+    Utils.assertRequiredParameterIsPresent(payload, 'payload');
+    Utils.assertRequiredParameterIsPresent(payload.type, 'payload.type');
+
+    this.payload = payload;
+    this.type = payload.type;
+    this.description = payload.description || null;
+
```
Comment (dbenchi) [inline lib/errors/JmapError.js:30]: why newline here?

File: lib/models/SetResponse.js [hunk 3]
```diff
@@ -55,4 +56,16 @@ export default class SetResponse extends Model {
 
     return new SetResponse(jmap, object);
   }
+
+  static _mapSetErrorsToJmapErrors(errors) {
+    let newErrors = {};
+
+    for (let key in errors) {
```
Comment (dbenchi) [inline lib/models/SetResponse.js:63]: Reflect.ownKeys does the same

File: lib/models/SetResponse.js [hunk 4]
```diff
@@ -55,4 +56,16 @@ export default class SetResponse extends Model {
 
     return new SetResponse(jmap, object);
   }
+
+  static _mapSetErrorsToJmapErrors(errors) {
```
Comment (dbenchi) [inline lib/models/SetResponse.js:60]: If it is a static method, then why does it start by `_`?

File: lib/utils/Utils.js [hunk 5]
```diff
@@ -154,10 +157,11 @@ export default class Utils {
       throw new Error('Expected an array of exactly 1 array element as the JMAP response for a "' + request + '" request.');
     }
 
-    let expectedResponse = Utils._expectedResponseFor(request);
+    let response = data[0][0],
+        expectedResponse = Utils._expectedResponseFor(request);
 
-    if (expectedResponse && data[0][0] !== expectedResponse) {
-      throw new Error('Expected "' + expectedResponse + '" as the JMAP response for a "' + request + '" request, but got "' + data[0][0] + '".');
+    if (response !== ERROR && (expectedResponse && response !== expectedResponse)) {
```
Comment (dbenchi) [inline lib/utils/Utils.js:163]: what is the diff between
`(response !== ERROR && (expectedResponse && response !== expectedResponse`
and
`(response !== ERROR && response !== expectedResponse`

why do I need to verify that `expectedResponse!== undefined`?

File: lib/utils/Utils.js [hunk 6]
```diff
@@ -154,10 +157,11 @@ export default class Utils {
...
+      throw new Error('Expected "' + expectedResponse + '" as the JMAP response for a "' + request + '" request, but got "' + response + '".');
```
Comment (dbenchi) [inline lib/utils/Utils.js]: backtick makes it easier to read here, WDYT?

File: lib/models/SetResponse.js (hunk 3 above)
Comment (ddolcimascolo) [reply lib/models/SetResponse.js:63]: Yeah, but requires another babel plugin :(

File: lib/errors/JmapError.js (hunk 2 above)
Comment (ddolcimascolo) [reply lib/errors/JmapError.js:30]: Readability, two blocks for two handling of two different parameters

File: test/backend/transport/RequestTransport.js [hunk 7]
```diff
@@ -22,24 +22,41 @@ describe('The RequestTransport class', function() {
...
+        done();
+      });
     });
 
     it('should reject the promise when status code is not 200 nor 201', function(done) {
       mockery.registerMock('request', function(options, callback) {
         callback(null, {
           statusCode: 400
-        });
+        }, '{}');
```
Comment (abenchi) [inline test/backend/transport/RequestTransport.js]: why do we add this param?

File: lib/models/SetResponse.js (hunk 4 above)
Comment (ddolcimascolo) [reply lib/models/SetResponse.js:60]: because it's static private :)

File: lib/utils/Utils.js (hunk 5 above)
Comment (ddolcimascolo) [reply lib/utils/Utils.js:163]: There's a test about it. For _forward_ compatibility, we do not want to throw if the backend sends us multiple responses with some responses we do not hande yet, like updates for instance.

File: test/common/Client.js [hunk 8]
```diff
@@ -2718,4 +2740,25 @@ describe('The Client class', function() {
 
   });
 
+  describe('When there is a JMAP error returned by the backend', function() {
+
+    it('should throw a JMAP error when an error is received from the backend', function(done) {
+      var client = defaultClient();
+
+      client.transport.post = function() {
+        return q([['error', { type: 'invalidArguments', description: 'The `id` property must be `singleton`.' }, '#0']]);
```
Comment (dbenchi) [inline test/common/Client.js:2753]: Just to refresh my memory: why do we pass nested array inside q? the question is why do we leverage data[0][0]

Comment (ddolcimascolo) [reply test/common/Client.js:2753]: It's the JMAP format.
A response can contain multiple business responses.

File: lib/utils/Utils.js (hunk 6 above)
Comment (ddolcimascolo) [reply lib/utils/Utils.js]: Changed

File: lib/client/Client.js (hunk 1 above)
Comment (ddolcimascolo) [reply lib/client/Client.js]: Changed

File: lib/models/SetResponse.js (hunk 3 above)
Comment (dbenchi) [reply lib/models/SetResponse.js:63]: FU babel.. I really hate it

**Pull request merged**